File: programmers/lv2/양궁 대회/양궁 대회.js
// https://school.programmers.co.kr/learn/courses/30/lessons/92342

// 라이언이 이기면 어피치랑 점수 차를 return
// 라이언이 지거나 비기면 -1 return
const getWhoWinWithDiff = (ryan, apeach) => {
  let ryanSum = 0;
  let apeachSum = 0;
  for (let i = 0; i < ryan.length; i++) {
    if (ryan[i] === apeach[i] && ryan[i] !== 0) {
      apeachSum += 10 - i;
    } else if (ryan[i] > apeach[i]) {
      ryanSum += 10 - i;
    } else if (ryan[i] < apeach[i]) {
      apeachSum += 10 - i;
    }
  }
  if (ryanSum > apeachSum) {
    return ryanSum - apeachSum;
  } else {
    return -1;
  }
};
const solution = (n, info) => {
  let q = [];
  let result = [];
  let maxDiff = -1;
  q.push([n, 0, Array(11).fill(0)]);
  while (q.length) {
    const [leftArrow, idx, ryanInfo] = q.shift();
    // 남은 화살이 없는 경우
    if (leftArrow === 0) {
      if (getWhoWinWithDiff(ryanInfo, info) > maxDiff) {
        maxDiff = getWhoWinWithDiff(ryanInfo, info);
        result = ryanInfo;
      } else if (getWhoWinWithDiff(ryanInfo, info) === maxDiff) {
        let isNewPossible = false;
        for (let i = 10; i >= 0; i--) {
          if (ryanInfo[i] > result[i]) {
            isNewPossible = true;
            break;
          }
        }
        result = isNewPossible && ryanInfo;
      }
    } // 마지막 쏘는 데 화살이 남은 경우
    else if (idx === 10) {
      let lastArray = [...ryanInfo];
      lastArray[10] = leftArrow;
      if (getWhoWinWithDiff(lastArray, info) > maxDiff) {
        maxDiff = getWhoWinWithDiff(lastArray, info);
        result = lastArray;
      } else if (getWhoWinWithDiff(ryanInfo, info) === maxDiff) {
        let isNewPossible = false;
        for (let i = 10; i >= 0; i--) {
          if (lastArray[i] > result[i]) {
            isNewPossible = true;
            break;
          }
        }
        result = isNewPossible && lastArray;
      }
    } else {
      // 라이언이 아예 안쏘는 경우
      q.push([leftArrow, idx + 1, ryanInfo]);
      // 라이언이 화살 쏘는 경우
      if (info[idx] + 1 <= leftArrow) {
        let array = [...ryanInfo];
        array[idx] = info[idx] + 1;
        q.push([leftArrow - (info[idx] + 1), idx + 1, array]);
      }
    }
  }
  return result ? result : [-1];
};
const info = [0, 0, 0, 0, 0, 0, 0, 0, 3, 4, 3];
const n = 10;
console.log(solution(n, info));

File: programmers/lv2/양궁 대회/양궁 대회.py
from collections import deque

def get_who_win_with_diff(ryan, apeach):
    ryan_sum = 0
    apeach_sum = 0
    for i in range(len(ryan)):
        if ryan[i] == apeach[i] and ryan[i] != 0:
            apeach_sum += 10 - i
        elif ryan[i] > apeach[i]:
            ryan_sum += 10 - i
        elif ryan[i] < apeach[i]:
            apeach_sum += 10 - i
    
  
    return ryan_sum - apeach_sum if ryan_sum > apeach_sum else -1

def solution(n, info):
    q = deque()
    result = []
    max_diff = -1
    q.append([n, 0, [0] * 11])
    
    while q:
        left_arrow, idx, ryan_info = q.popleft()
        
        # 남은 화살이 없는 경우
        if left_arrow == 0:
            diff = get_who_win_with_diff(ryan_info, info)
            if diff > max_diff:
                max_diff = diff
                result = ryan_info[:]
            elif diff == max_diff:
                is_new_possible = False
                for i in range(10, -1, -1):
                    if ryan_info[i] > result[i]:
                        is_new_possible = True
                        break
                if is_new_possible:
                    result = ryan_info[:]
        
        # 마지막 쏘는 데 화살이 남은 경우
        elif idx == 10:
            last_array = ryan_info[:]
            last_array[10] = left_arrow
            diff = get_who_win_with_diff(last_array, info)
            if diff > max_diff:
                max_diff = diff
                result = last_array[:]
            elif diff == max_diff:
                is_new_possible = False
                for i in range(10, -1, -1):
                    if last_array[i] > result[i]:
                        is_new_possible = True
                        break
                if is_new_possible:
                    result = last_array[:]
        
        else:
            # 라이언이 아예 안쏘는 경우
            q.append([left_arrow, idx + 1, ryan_info[:]])
            
            # 라이언이 화살 쏘는 경우
            if info[idx] + 1 <= left_arrow:
                array = ryan_info[:]
                array[idx] = info[idx] + 1
                q.append([left_arrow - (info[idx] + 1), idx + 1, array])
    
    return result if result else [-1]

info = [0, 0, 0, 0, 0, 0, 0, 0, 3, 4, 3]
n = 10
print(solution(n, info))